docs/implementation.adoc

```diff
@@ -0,0 +1,507 @@
+
+NOTE: When implementing categorization it is recommended to reuse an existing categorization class (with the terminology injected) rather than create a new categorization.
+New categorizations should be added to the scm-api plugin by pull requests as this allows similar categories to be grouped.
+
+Branding controls the visual icons that are used to represent the `jenkins.scm.api.SCMSource` and `jenkins.scm.api.SCMNavigator` instances.
+Branding is determined by the `getIconClassName()` of the `jenkins.scm.api.SCMSourceDescriptor` and `jenkins.scm.api.SCMNavigatorDescriptor`.
+Where these methods return non-null the corresponding icons will be used by consumers of the SCM API as the final fall-back icons.
+
+=== Implementing `hudson.scm.SCM`
+
+The `hudson.scm.SCM` API has been subject to significant evolution. Modern implementations should focus on implementing the following methods:
+
+[source,java]
+----
+public class MySCM extends SCM {
+    /*
+     * all configuration fields should be private
+     * mandatory fields should be final
+     * non-mandatory fields should be non-final
+     */
+
+    @DataBoundConstructor
+    public MySCM(/*mandatory configuration*/) {
+        // ...
+    }
+
+    // for easier interop with SCMSource
+    public MySCM(MySCMSource config) {
+        // copy the configuratuion from the SCMSource
+    }
+
+    // Getters for all the configuration fields
+
+    // use @DataBoundSetter to inject the non-mandatory configuration elements
+    // as this will simplify the usage from pipeline
+
+    @Override
+    public boolean supportsPolling() {
+        return true; // hopefully you do
+    }
+
+    @Override
+    public boolean requiresWorkspaceForPolling() {
+        return false; // hopefully you don't
+    }
+
+    // for easier interop with SCMSource
+    public void setSCMHead(@NonNull SCMHead head, @CheckForNull SCMRevision revision) {
+        // configure to checkout the specified head at the specific revision
+        // if passed implementations that do not come from a MySCMSource then silently ignore
+    }
+
+    @Override
+    public PollingResult compareRemoteRevisionWith(@Nonnull Job<?, ?> project, @Nullable Launcher launcher,
+                                                   @Nullable FilePath workspace, @Nonnull TaskListener listener,
+                                                   @Nonnull SCMRevisionState baseline)
+            throws IOException, InterruptedException {
+        if (baseline instanceof MySCMRevisionState) {
+            //
+            // get current revision in SCM
+            // if your implementation of requiresWorkspaceForPolling() returns true then the
+            // workspace and launcher parameters should be non-null and point to a
+            // workspace and node to use for the comparison
+            // NOTE: requiring a workspace for polling is a realy bad user experience
```

Comment:
We should fix the Git plugin to use the local cache when path exclusions etc. are configured…